output-comparator/src/main/java/org/mobilitydata/gtfsvalidator/outputcomparator/io/DatasetMemoryUsage.java

```diff
@@ -0,0 +1,72 @@
+package org.mobilitydata.gtfsvalidator.outputcomparator.io;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.mobilitydata.gtfsvalidator.performance.MemoryUsage;
+
+/**
+ * Represents memory usage information for a dataset. This class contains the information associated
+ * with the memory usage of a dataset when running the validation process.
+ */
+public class DatasetMemoryUsage {
+
+  private String datasetId;
+  private List<MemoryUsage> referenceMemoryUsage;
+  private List<MemoryUsage> latestMemoryUsage;
+  private Map<String, Long> referenceUsedMemoryByKey = Collections.unmodifiableMap(new HashMap<>());
+  private Map<String, Long> latestUsedMemoryByKey = Collections.unmodifiableMap(new HashMap<>());
+
+  public DatasetMemoryUsage(
+      String datasetId,
+      List<MemoryUsage> referenceMemoryUsage,
+      List<MemoryUsage> latestMemoryUsage) {
+    this.datasetId = datasetId;
+    this.referenceMemoryUsage = referenceMemoryUsage;
+    this.latestMemoryUsage = latestMemoryUsage;
+    if (referenceMemoryUsage != null) {
+      this.referenceUsedMemoryByKey =
```

Comment:
What happens if the list contains the key twice? Can it happen?

Reply:
Yes, it can happen. When the annotation is used, it's less likely to happen as the key is the complete method signature, including the class name and path. For human-created keys, it's possible. I don't know if adding an exception or interrupting the process is a good idea, as extracting the metrics is a non-essential operation. Logging a warning can be buried within the GitHub action. Do you have any suggestions? _Other option is forcing uniqueness before adding the key to the list, but I landed in the same place, should we interrupt the validation process?_

Reply:
I think it can happen with the comparator when reading a json report file? Please confirm. 
We expect the file to have been written properly by the program, but it's true that a user could modify this file, or there could be a bug when the file is written by the validator. 
I think that just ignoring the second element with the same key with a log could be the way to go. It's a corner case IMO so just a message in the logs should be OK. To me it's more important the the possible exception does not stop the whole process.
I think to do that you'll have to use an explicit loop instead of streams, right?

Reply:
Yes, this is when the comparator reads the JSON report. Theoretically, the list in the report can have duplicates, so duplicates are still properly defined in a JSON report. In practice, we are currently adding only three keys. I'll add a log. For clarification, there is no exception thrown if duplicates are found in the report, we are just ignoring/overriding the values.

Reply:
A bit late, but I was revisiting this and with the way toUnmodifiableMap was used before (toUnmodifiableMap(MemoryUsage::getKey, MemoryUsage::usedMemory)) there was a risk of a IllegalStateException according to [this](https://docs.oracle.com/javase/10/docs/api/java/util/stream/Collectors.html#toUnmodifiableMap(java.util.function.Function,java.util.function.Function))
The 3 parameters function that you used in the end does not throw that exception, so all is well.